### PR TITLE
Add SetPreSignatureDecoder

### PR DESCRIPTION
**Pull request by fmrsabino**
Adds the `SetPreSignatureDecoder`, a decoder helper which can be used to retrieve a CoW Swap Order UID from any `setPreSignature` transaction.

This represent preliminary work for https://github.com/safe-global/safe-client-gateway/issues/1198. With the ability to retrieve the Order UID (this PR) and the integration of the CoW Swap API done in https://github.com/safe-global/safe-client-gateway/pull/1233, the CGW components can now retrieve the data of CoW Swap orders and map the respective data to be rendered to the clients.

### CODE REVIEW

File: src/domain/swaps/contracts/decoders/set-pre-signature-decoder.helper.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  /**
+   * Gets the Order UID associated with the provided transaction data.
+   *
+   * @param data - the transaction data for the setPreSignature call
+   * @returns {`0x${string}`} the order UID or null if the data does not represent a setPreSignature transaction
+   */
+  getOrderUid(data: `0x${string}`): `0x${string}` | null {
+    if (!data.startsWith(this.setPreSignatureFunctionSelector)) return null;
+    const { functionName, args } = this.decodeFunctionData({ data });
+    if (functionName !== 'setPreSignature') return null;
```
Comment (iamacook) [inline src/domain/swaps/contracts/decoders/set-pre-signature-decoder.helper.ts:27]: Line 25 and 27 check the same thing here. If the former is removed, `decodeFunctionData` can throw though.

File: src/domain/swaps/contracts/decoders/set-pre-signature-decoder.helper.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    if (!args || !args[0] || typeof args[0] !== 'string' || !isHex(args[0]))
```
Comment (iamacook) [inline src/domain/swaps/contracts/decoders/set-pre-signature-decoder.helper.ts:28]: These checks are theoretically not needed on a type level.

Are you checking for "fake" signatures? If so, you don't need to check `args[0]` or the `typeof` as the arguments are always an array (if present) and `isHex` accepts an `unknown` type.

File: src/domain/swaps/contracts/decoders/set-pre-signature-decoder.helper.ts (hunk 1 above)
Comment (fmrsabino) [reply src/domain/swaps/contracts/decoders/set-pre-signature-decoder.helper.ts:27]: Good call about the `throw` (honestly I think this should be documented). I will wrap it in a `try/catch` block and return `null`. The goal of this function is not to provide the caller the opportunity to deal with the error.

File: src/domain/swaps/contracts/decoders/set-pre-signature-decoder.helper.ts (hunk 2 above)
Comment (fmrsabino) [reply src/domain/swaps/contracts/decoders/set-pre-signature-decoder.helper.ts]: Right! At some point I had a state where the type checks were not working but `decodeFunctionData` does return the correct types. Fixed in [9dca8df](https://github.com/safe-global/safe-client-gateway/pull/1239/commits/9dca8df8a3cde5031ea9fd75fc93e15fb4a30639)

File: src/domain/swaps/contracts/decoders/set-pre-signature-decoder.helper.ts (hunk 1 above)
Comment (fmrsabino) [reply src/domain/swaps/contracts/decoders/set-pre-signature-decoder.helper.ts]: Fixed in [9dca8df](https://github.com/safe-global/safe-client-gateway/pull/1239/commits/9dca8df8a3cde5031ea9fd75fc93e15fb4a30639)

**Review by iamacook — APPROVED** [review]

**Pull request merged**